File: services/parser-rs/src/core/pdf_parser.rs
use crate::core::layout_modeler::LayoutModeler;
use crate::{DocumentSection, ElementType};
use anyhow::Result;
use std::collections::HashMap;
use std::path::Path;

pub trait PdfParser {
    /// Parse a PDF document and return a DocumentTree
    fn parse<P: AsRef<Path>>(&self, path: P) -> Result<crate::core::layout::DocumentTree>;
}

/// PdfParser implementation using lopdf
pub struct LopdfParser;

impl PdfParser for LopdfParser {
    fn parse<P: AsRef<Path>>(&self, path: P) -> Result<crate::core::layout::DocumentTree> {
        use lopdf::Document;

        // Load the PDF document
        let doc = Document::load(path.as_ref())?;

        let mut sections = Vec::new();
        let mut section_id = 1;

        // Get all page numbers
        let pages = doc.get_pages();

        // Iterate through pages
        for (page_num, _) in pages.iter() {
            // Extract text from the page
            if let Ok(content) = doc.extract_text(&[*page_num]) {
                // Split content into paragraphs/sections
                let paragraphs: Vec<&str> = content.lines().collect();

                for paragraph in paragraphs {
                    if !paragraph.trim().is_empty() {
                        let formatting = self.extract_formatting_for_pdf(paragraph);

                        let section = DocumentSection {
                            id: section_id,
                            element_type: ElementType::Paragraph, // For now, treat everything as paragraph
                            raw_text: paragraph.to_string(),
                            formatting,
                            xml_path: format!("page_{}.txt", page_num), // Using page number as reference
                        };

                        sections.push(section);
                        section_id += 1;
                    }
                }
            }
        }

        let mut modeler = LayoutModeler::new();
        let mut document_tree = modeler.build_tree(sections);
        document_tree.metadata.total_pages = Some(pages.len() as u32);
        document_tree.metadata.file_path = Some(path.as_ref().to_string_lossy().to_string());
        document_tree.metadata.file_size = Some(std::fs::metadata(path.as_ref())?.len());

        Ok(document_tree)
    }
}

impl LopdfParser {
    /// Extract formatting information from PDF text
    fn extract_formatting_for_pdf(&self, _text: &str) -> HashMap<String, String> {
        let mut formatting = HashMap::new();
        formatting.insert("element-type".to_string(), "pdf-text".to_string());
        formatting
    }
}

File: services/parser-py/src/core/pdf_parser.py
import PyPDF2
from pathlib import Path
from typing import Dict, Any

class PdfParser:
    def parse(self, path: Path) -> Dict[str, Any]:
        sections = []
        positions = {}
        section_id = 1
        
        with open(path, 'rb') as f:
            reader = PyPDF2.PdfReader(f)
            num_pages = len(reader.pages)
            
            for page_num in range(num_pages):
                page = reader.pages[page_num]
                text = page.extract_text()
                
                if text:
                    paragraphs = text.split('\n')
                    for para_index, paragraph in enumerate(paragraphs):
                        if paragraph.strip():
                            # 确保每个元素都包含格式信息
                            section = {
                                'id': section_id,
                                'element_type': 'Paragraph',
                                'raw_text': paragraph.strip(),
                                'formatting': {
                                    'element-type': 'pdf-text',
                                    'font-size': '12pt',  # 默认字体大小
                                    'font-family': 'Arial',  # 默认字体类型
                                    'font-weight': 'normal'  # 默认字体粗细
                                },
                                'xml_path': f'page_{page_num + 1}_para_{para_index + 1}.txt',
                                'offset': id(paragraph)  # 添加偏移量
                            }
                            
                            # 生成位置信息
                            position = {
                                'x': 50.0,
                                'y': 50.0 + (para_index * 20),
                                'width': 500.0,
                                'height': 15.0,
                                'page_number': page_num + 1
                            }
                            
                            sections.append(section)
                            positions[section_id] = position
                            section_id += 1
        
        metadata = {
            'title': None,
            'total_pages': len(reader.pages) if 'reader' in locals() else 0,
            'file_path': str(path),
            'file_size': path.stat().st_size,
            'creation_date': None,
            'modification_date': None
        }
        
        return {
            'sections': sections,
            'positions': positions,
            'metadata': metadata
        }
